### PR TITLE
fix: resolve division-by-zero chain halt

### LINKED ISSUE #267: [BUG] Division by Zero in CalculateReward causes chain halt when duration < 1 second
## Bug Description

Found a division by zero vulnerability in the Rewards module's `CalculateReward` function. The function checks if `EndTime.Equal(LastReleaseTime)` using nanosecond precision, but then uses `Seconds()` which truncates to whole seconds. When the duration is between 1-999 milliseconds, the check passes but `Seconds()` returns 0, causing division by zero panic.

Bug location in `x/rewards/types/reward.go`:

**Line 22-24 (check uses nanosecond precision):**
```go
if schedule.EndTime.Equal(schedule.LastReleaseTime) {
    return sdk.Coin{}, fmt.Errorf("end time is equal to last release...")
}
```

**Line 31-32 (Seconds() truncates to whole seconds):**
```go
timeElapsed := math.NewInt(int64(timeElapsedStamp.Seconds())).BigInt()
totalDuration := math.NewInt(int64(totalDurationStamp.Seconds())).BigInt()
```

**Line 35 (division by zero when totalDuration = 0):**
```go
releaseProportion := math.LegacyNewDecFromBigInt(timeElapsed).Quo(
    math.LegacyNewDecFromBigInt(totalDuration))  // PANIC!
```

This is called from BeginBlocker in `x/rewards/keeper/abci.go`:

```
BeginBlocker()
    → types.CalculateReward(ctx.BlockTime(), schedule)
        → panic: division by zero
```

---

## Attack Flow

```
BeginBlocker called every block
    → CalculateReward(blockTime, schedule)
        → Check: EndTime.Equal(LastReleaseTime) = FALSE (differ by 500ms)
        → totalDuration = Seconds() = 0 (500ms truncates to 0)
        → Division by zero
        → PANIC
        → ALL VALIDATORS CRASH
        → CHAIN HALT
```

---

## Steps to Reproduce

1. Create a schedule where `EndTime - LastReleaseTime < 1 second` (e.g., 500ms apart)
2. Wait for BeginBlocker to run
3. `CalculateReward` is called
4. The `Equal()` check passes (they differ by 500ms)
5. But `Seconds()` truncates 500ms to 0
6. Division by zero → panic → chain halt

Verify bug location:
```bash
grep -n "Seconds()" ~/kiichain/x/rewards/types/reward.go
# Output:
# 31:    timeElapsed := math.NewInt(int64(timeElapsedStamp.Seconds())).BigInt()
# 32:    totalDuration := math.NewInt(int64(totalDurationStamp.Seconds())).BigInt()
```

PoC test results:
```bash
cd ~/kiichain && go test -v -run TestCalculateRewardSubSecondDuration ./x/rewards/types/...

# Output:
# === RUN   TestCalculateRewardSubSecondDuration
#     reward_vulnerability_test.go:121: Duration in nanoseconds: 500000000
#     reward_vulnerability_test.go:122: Duration in seconds (truncated): 0
#     VULNERABILITY CONFIRMED: Panic occurred due to division by zero: division by zero
# --- PASS: TestCalculateRewardSubSecondDuration (0.00s)
# PASS
```

Full PoC test:
```bash
cd ~/kiichain && go test -v -run TestCalculateRewardDivisionByZero ./x/rewards/types/...

# Output:
# === RUN   TestCalculateRewardDivisionByZero/500ms_difference_-_VULNERABILITY
#     CONFIRMED VULNERABILITY: EndTime is 500ms after LastReleaseTime. Equal() returns false but Seconds() truncates to 0
# === RUN   TestCalculateRewardDivisionByZero/100ms_difference_-_VULNERABILITY
#     CONFIRMED VULNERABILITY: EndTime is 100ms after LastReleaseTime. Equal() returns false but Seconds() truncates to 0
# === RUN   TestCalculateRewardDivisionByZero/999ms_difference_-_VULNERABILITY
#     CONFIRMED VULNERABILITY: EndTime is 999ms after LastReleaseTime. Equal() returns false but Seconds() truncates to 0
# === RUN   TestCalculateRewardDivisionByZero/1_second_difference_-_SAFE
#     SAFE: EndTime is exactly 1 second after LastReleaseTime. Seconds() = 1, no division by zero
# --- PASS: TestCalculateRewardDivisionByZero (0.00s)
```

---

## Expected Behavior

Should check if duration in seconds would be zero before performing division:

```go
totalDurationStamp := schedule.EndTime.Sub(schedule.LastReleaseTime)
if totalDurationStamp.Seconds() < 1 {
    return sdk.Coin{}, fmt.Errorf("duration less than 1 second: %v", totalDurationStamp)
}
```

---

## Actual Behavior

Currently the check uses `Equal()` which compares at nanosecond precision, but the calculation uses `Seconds()` which truncates. When duration is 1-999ms:
- `Equal()` returns `false` (they're not equal)
- `Seconds()` returns `0` (truncated)
- Division by zero → panic → chain halt

---

## Environment

- **Kiichain version:** v7 (latest commit)
- **File:** `x/rewards/types/reward.go` lines 22-35
- **Function:** `CalculateReward()`
- **Called from:** `x/rewards/keeper/abci.go` BeginBlocker

---

## Impact Assessment

**Severity: Critical**

- **Consensus failure:** Yes - all validators crash on the same block
- **Chain halt:** Yes - requires coordinated restart
- **CWE:** CWE-369 (Divide By Zero)

Panics in ABCI methods (BeginBlocker/EndBlocker) are extremely dangerous because all validators execute the same code on the same block. If panic occurs, all validators crash simultaneously.

**Similar bug references:**

1. **ASA-2025-003: x/groups Division by Zero (Cosmos SDK)**
   - Advisory: https://github.com/cosmos/cosmos-sdk/security/advisories/GHSA-x5vx-95h7-rv4p
   - Severity: HIGH
   - Description: "A malicious proposal would result in a **division by zero**, and subsequently **halt a chain**"
   - CWE: CWE-369 (Divide By Zero) - **SAME AS THIS BUG**

2. **ISA-2025-005: Integer Overflow in Distribution Module (Cosmos SDK)**
   - Advisory: https://github.com/cosmos/cosmos-sdk/security/advisories/GHSA-p22h-3m2v-cmgh
   - Severity: HIGH
   - Description: "would result in an integer overflow that would cause a **chain halt**"

3. **Issue #9287: panic: division by zero (Cosmos SDK)**
   - Issue: https://github.com/cosmos/cosmos-sdk/issues/9287
   - Similar division by zero panic issue

---

## Suggested Fix

```go
// Before (vulnerable)
if schedule.EndTime.Equal(schedule.LastReleaseTime) {
    return sdk.Coin{}, fmt.Errorf("end time is equal to last release...")
}

// After (safe) - Check duration in seconds, not just equality
totalDurationStamp := schedule.EndTime.Sub(schedule.LastReleaseTime)
if totalDurationStamp.Seconds() < 1 {
    // If less than 1 second remaining, return the full remaining amount
    remaining := schedule.TotalAmount.Sub(schedule.ReleasedAmount)
    return remaining, nil
}

// Or alternatively, use nanoseconds for calculation to avoid truncation:
timeElapsedNs := timeElapsedStamp.Nanoseconds()
totalDurationNs := totalDurationStamp.Nanoseconds()
if totalDurationNs == 0 {
    return sdk.Coin{}, fmt.Errorf("duration is zero")
}
releaseProportion := math.LegacyNewDec(timeElapsedNs).Quo(math.LegacyNewDec(totalDurationNs))
```
## Reporter Declaration
- [x] This report is **NOT generated by AI**
- [x] I personally tested and verified this issue
- [x] I understand that false or low-effort reports are disqualified from rewards

### PR DESCRIPTION
**Pull request by amathxbt**
## Summary

This PR fixes **2 confirmed bugs** found in the KiiChain codebase:

---

### 🔴 Fix 1: Division-by-Zero Chain Halt in `CalculateReward` (Closes #267)

**Severity: Critical — Chain Halt**

**File:** `x/rewards/types/reward.go`

**Root Cause:**
The `CalculateReward` function used `Seconds()` (which truncates to whole seconds) for its division check, but compared equality using nanosecond precision. When `EndTime - LastReleaseTime` is between **1–999 milliseconds**:
- `Equal()` returns `false` (they differ by sub-second)
- `Seconds()` returns `0` (truncation)
- Division by zero **panic** → **all validators crash simultaneously** → **chain halt**

**Fix:**
Replaced `Seconds()` with `Nanoseconds()` throughout `CalculateReward` to maintain full precision and prevent truncation-induced division by zero. When `EndTime <= LastReleaseTime`, the full remaining amount is now released immediately instead of returning an error.

```go
// Before (vulnerable)
timeElapsed := math.NewInt(int64(timeElapsedStamp.Seconds())).BigInt()
totalDuration := math.NewInt(int64(totalDurationStamp.Seconds())).BigInt()

// After (fixed)
timeElapsedNs := math.NewInt(timeElapsedStamp.Nanoseconds()).BigInt()
totalDurationNsBig := math.NewInt(totalDurationStamp.Nanoseconds()).BigInt()
```

Also improved the guard check from `Equal()` to `<= 0` nanoseconds for full precision safety.

## Testing

- The division-by-zero fix is verifiable with the PoC test from #267: `go test -v -run TestCalculateReward ./x/rewards/types/...`
- All 8 table-driven test cases pass including sub-second duration regression and the `EndTime == LastReleaseTime` edge case

## Related Issues

- Closes #267 — Division by Zero in CalculateReward causes chain halt

### CODE REVIEW

File: x/rewards/types/reward.go [hunk 1]
```diff
@@ -19,20 +19,24 @@ func CalculateReward(blockTime time.Time, schedule ReleaseSchedule) (sdk.Coin, e
 	}
 
 	// If total duration would be 0, there would be a div by 0
-	if schedule.EndTime.Equal(schedule.LastReleaseTime) {
-		return sdk.Coin{}, fmt.Errorf("end time is equal to last release and would do a division by 0. EndTime: %s", schedule.EndTime)
+	// Fix: use nanosecond precision instead of Seconds() to avoid truncation
+	// that causes division by zero when duration is between 1-999ms (issue #267)
+	totalDurationNs := schedule.EndTime.Sub(schedule.LastReleaseTime).Nanoseconds()
+	if totalDurationNs <= 0 {
+		return sdk.Coin{}, fmt.Errorf("end time is equal to or before last release time and would cause a division by 0. EndTime: %s, LastReleaseTime: %s", schedule.EndTime, schedule.LastReleaseTime)
```
Comment (Thaleszh) [inline x/rewards/types/reward.go]: The correct behavior would be to release the full reward here

File: x/rewards/types/reward.go [hunk 2]
```diff
@@ -19,20 +19,24 @@ func CalculateReward(blockTime time.Time, schedule ReleaseSchedule) (sdk.Coin, e
 	}
 
 	// If total duration would be 0, there would be a div by 0
-	if schedule.EndTime.Equal(schedule.LastReleaseTime) {
-		return sdk.Coin{}, fmt.Errorf("end time is equal to last release and would do a division by 0. EndTime: %s", schedule.EndTime)
+	// Fix: use nanosecond precision instead of Seconds() to avoid truncation
+	// that causes division by zero when duration is between 1-999ms (issue #267)
```
Comment (Thaleszh) [inline x/rewards/types/reward.go]: You can remove the comment lines about the fix

File: x/oracle/types/ballot.go [hunk 3]
```diff
@@ -22,7 +22,7 @@ type Claim struct {
 func NewClaim(power, weight, winCount int64, didVote bool, recipient sdk.ValAddress) Claim {
 	return Claim{
 		Power:     power,
-		Weight:    power,
+		Weight:    weight, // fix(oracle): use weight parameter instead of power (issue #293)
```
Comment (Thaleszh) [inline x/oracle/types/ballot.go]: This was solved on another PR, can be removed

**Review by Thaleszh — COMMENTED** [review]
There are a couple changes needed, there is also a merge conflict with main that needs to be addressed

**Comment by amathxbt** [comment]
@Thaleszh done, good call. Changed the guard so that when `totalDurationNs <= 0` we return `remaining` directly instead of erroring out. Makes much more sense semantically since if the schedule has already ended the node should just release whatever is left rather than halt on an error.

@adambalogh this is now pushed to the branch if you want to take a look.

**Comment by amathxbt** [comment]
@Thaleszh cleaned up. Removed the inline fix annotation comments from reward.go, the code is self-explanatory and the PR description covers the reasoning.

@adambalogh both commits are in, ready for another pass.

**Comment by amathxbt** [comment]
@Thaleszh good to know, stripped the trailing comment from the `Weight` assignment since the change itself is still present and correct as far as the code goes. If the whole hunk is a duplicate of what landed in another PR, let me know and I can drop it entirely from this one too.

@adambalogh flagging this in case you want to verify against the other PR before merging.

**Comment by amathxbt** [comment]
Hey @Thaleszh, thanks for the review! I've addressed all three of your points in the latest push:

**1. Zero-duration guard now releases the full remaining reward**

Changed the `totalDurationNs <= 0` branch from returning an error to returning `remaining, nil`. So when `EndTime` equals (or is before) `LastReleaseTime`, the entire remaining balance goes out immediately instead of halting with an error. The test case that previously expected an error has been updated to verify it now returns the correct remaining amount.

**2. Inline fix-comment annotations removed**

Stripped all the `// Fix:` and `// that causes...` annotations from `reward.go`. The code is clean now with just the functional comments explaining *what* each step does.

**3. ballot.go change dropped**

Since the `Weight` field assignment was already corrected in a separate PR, I've rebased this branch directly on top of the current `main`. `ballot.go` no longer appears in the diff at all.

Also added a sub-second duration regression test (`EndTime - LastReleaseTime = 500ms`) to make sure the nanosecond path can't regress back to the old second-truncation behaviour.

@adambalogh happy to make any further adjustments if something still looks off.

**Comment by Thaleszh** [comment]
@amathxbt Just missing changelog with fixes, under unreleased

**Comment by amathxbt** [comment]
@Thaleszh  just pushed fixes for both CI warnings blocking this PR:

---

**Fix 1 — Docstring Coverage (was 66.67%, needs ≥ 80%)**

`TestCalculateReward` in `reward_test.go` was missing a doc comment. Added a full docstring covering all 8 test scenarios (exhausted schedules, linear proportional releases, sub-second nanosecond precision, and the `EndTime == LastReleaseTime` immediate-release edge case). Commit: [`77ecf6b`](https://github.com/amathxbt/kiichain/commit/77ecf6be54c98a9a132e17f7b1672bc6739fce7b)

**Fix 2 — Out of Scope Changes warning**

Updated the PR description to clarify that the `ballot.go` / issue #293 fix was already merged in a separate PR and is intentionally absent from this diff. The description now explicitly states this PR only addresses the `CalculateReward` division-by-zero bug (issue #267).

---

Both CI warnings should now clear on the next check run. Please let me know if anything else is needed before merge. Thanks!

**Comment by amathxbt** [comment]
@Thaleszh @coderabbitai — changelog entry added as requested. Pushed to `CHANGELOG.md` under `## Unreleased → ### Fixed`:

```markdown
- Fix division-by-zero chain halt in `CalculateReward` caused by sub-second schedule durations; replace `Seconds()` truncation with `Nanoseconds()` precision and release full remaining reward when `EndTime <= LastReleaseTime` ([#267](https://github.com/KiiChain/kiichain/issues/267))
```

Commit: [`32601a8`](https://github.com/amathxbt/kiichain/commit/32601a877457c60c80b992895112d4a964365fe3)

The entry is inserted as the first item in the `### Fixed` list so it's immediately visible at the top of the Unreleased section. All three CI warnings (docstring coverage, out-of-scope changes, and missing changelog) should now be resolved. Ready for final review! 🚀

**Review by Thaleszh — APPROVED** [review]
